### PR TITLE
feat(logging): enable log rotation and set retry on full log store sync

### PR DESCRIPTION
**Pull request by NikaHsn**
*Issue #, if available:*

*Description of changes:*
- make QueuedItemStore.isFull() sync
- enable log rotation when log store is full
- set retry on full log store sync
- handle rejectedLogEventsInfo response and do partial sync

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 1]
```diff
@@ -139,32 +142,110 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+            // set logs to start from next
+            logs.removeRange(0, tooOldLogEventEndIndex + 1);
+            // set events to start from next
+            events.removeRange(0, tooOldLogEventEndIndex + 1);
+            continue;
+          }
+          if (tooNewLogEventStartIndex != null &&
+              tooNewLogEventStartIndex >= 0 &&
+              tooNewLogEventStartIndex <= events.length - 1) {
+            tooNewException = _TooNewLogEventException(
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:183]: Why not throw it here? The control flow would be easier to follow

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 2]
```diff
@@ -139,32 +142,110 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+        _handleFullLogStoreAfterSync(
+          retryTimeInMillisecondsSinceEpoch: nextRetry,
+        );
         _syncing = false;
       }
     }
   }
 
+  void _handleFullLogStoreAfterSync({
+    int retryTimeInMillisecondsSinceEpoch = 0,
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:221]: Can you change this logic to work with `DateTime` and `Duration` objects instead of integers? This is very opaque and hard to follow

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 3]
```diff
@@ -139,32 +142,110 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+              tooNewLogEventStartIndex >= 0 &&
+              tooNewLogEventStartIndex <= events.length - 1) {
+            tooNewException = _TooNewLogEventException(
+              events[tooNewLogEventStartIndex].timestamp.toInt(),
+            );
+            // set logs to end at the index
+            logs.removeRange(tooNewLogEventStartIndex, logs.length);
+            // set events to end at the index
+            events.removeRange(tooNewLogEventStartIndex, events.length);
+            continue;
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:190]: Why continue if all batches after this will be newer?

File: packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart [hunk 4]
```diff
@@ -167,15 +167,14 @@ class IndexedDbAdapter implements QueuedItemStore {
...
+  static bool checkIsIndexedDBSupported() {
     if (indexedDB == null) {
       return false;
     }
     // indexedDB will be non-null in Firefox private browsing,
     // but will fail to open.
     try {
-      final openRequest = indexedDB!.open('test', 1);
-      await openRequest.future;
+      indexedDB!.open('test', 1).result;
```
[the remark this answers is not in the document]
Comment (NikaHsn) [reply packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart:177]: `QueuedItemStore.isFull()` is a sync method. however it was implemented async because the web implementation calls `checkIsIndexedDBSupported` to either use indexedDB or InMemoryQueuedItemStore. Because the `checkIsIndexedDBSupported` was async all the web APIs had to be async.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 1 above)
Comment (NikaHsn) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:183]: we want to continue with syncing the current batch after sanitizing the events in the current batch and then throw to stop syncing next batches

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 3 above)
Comment (NikaHsn) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:190]: we want to continue with syncing the current batch after sanitizing the events in the current batch and then throw to stop syncing next batches

File: packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart:177]: But why would this be guaranteed to throw in the same way as before?

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 5]
```diff
@@ -295,3 +364,34 @@ extension on LogEntry {
     );
   }
 }
+
+extension on RejectedLogEventsInfo {
+  (int? pastEndIndex, int? futureStartIndex) parse(int length) {
+    int? pastEndIndex;
+    int? futureStartIndex;
+
+    if (_isValidIndex(tooOldLogEventEndIndex, length)) {
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:373]: Why is this allowed to be false? I would use `RangeError.checkValidIndex` to assert these are valid.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 6]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+              (await _sendToCloudWatch(events)).rejectedLogEventsInfo;
+          if (rejectedLogEventsInfo == null) {
+            await _logStore.deleteItems(logs);
+            break;
+          }
+          {
+            final (tooOldEndIndex, tooNewStartIndex) =
+                rejectedLogEventsInfo.parse(events.length);
+
+            if (_isValidIndex(tooNewStartIndex, events.length)) {
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:156]: You've already checked this

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 7]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
-          break;
+        _TooNewLogEventException? tooNewException;
+        while (logs.isNotEmpty && events.isNotEmpty) {
+          final rejectedLogEventsInfo =
+              (await _sendToCloudWatch(events)).rejectedLogEventsInfo;
+          if (rejectedLogEventsInfo == null) {
+            await _logStore.deleteItems(logs);
+            break;
+          }
+          {
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:152]: Remove these braces; they're unnecessary

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 8]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+            }
+            if (_isValidIndex(tooOldEndIndex, events.length)) {
+              // remove old logs from log store.
+              await _logStore.deleteItems(logs.sublist(0, tooOldEndIndex! + 1));
+              // set logs to start after the index.
+              logs.removeRange(0, tooOldEndIndex + 1);
+              // set events to start after the index.
+              events.removeRange(0, tooOldEndIndex + 1);
+            }
+            continue;
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:173]: Can remove `continue` now

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 9]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
   }
 
+  void _handleFullLogStoreAfterSync({
+    DateTime? retryTime,
+  }) {
+    final isLogStoreFull =
+        _logStore.isFull(_pluginConfig.localStoreMaxSizeInMB);
+    if (!isLogStoreFull) {
+      _retryCount = 0;
+      _retryTime = null;
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:211]: Why would you not respect `retryTime` here?

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 10]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+      return;
+    }
+    if (retryTime != null && retryTime.isAfter(DateTime.timestamp())) {
+      _retryTime = retryTime;
+      return;
+    }
+    _retryCount += 1;
+    _retryTime = DateTime.timestamp().add((_baseRetryInterval * _retryCount));
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:219]: This seems like a long time for a basic backoff

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 11]
```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+              // set events to start after the index.
+              events.removeRange(0, tooOldEndIndex + 1);
+            }
+            continue;
+          }
+        }
+        // after sending each batch to CloudWatch check if the batch has
+        // `tooNewException` and throw to stop syncing next batches.
+        if (tooNewException != null) {
+          throw tooNewException;
```
Comment (dnys1) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:179]: I think this would be cleaner if you just returned the retry time instead of throwing

File: packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (NikaHsn) [reply packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart:177]: based on [the docs ](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/open) If the operation is successful, the value of result is a connection to the database. If the request failed and the result is not available, an InvalidStateError exception is thrown.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 10 above)
Comment (NikaHsn) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:217]: `flushIntervalInSeconds = const Duration(seconds: 60)`. updated the `_baseRetryInterval` to 10 secs so there will be 2-3 retries within the flush interval.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 9 above)
Comment (NikaHsn) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:209]: it is reseting retry if log store is not full. it retries to sync only if log store is full otherwise can wait till next sync.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 6 above)
Comment (NikaHsn) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart]: `RejectedLogEventsInfo.parse()` validates RejectedLogEventsInfo indices. 
here is validating the indices returned by parse() method.

**Review by dnys1 — APPROVED** [review]
